zk_dtypes/include/field/quartic_extension_field_operation.h

```diff
@@ -66,7 +66,89 @@ class QuarticExtensionFieldOperation : public ExtensionFieldOperation<Derived>,
   Derived Square() const {
     ExtensionFieldMulAlgorithm algorithm =
         static_cast<const Derived&>(*this).GetSquareAlgorithm();
-    if (algorithm == ExtensionFieldMulAlgorithm::kToomCook) {
+    if (algorithm == ExtensionFieldMulAlgorithm::kCustom) {
+      // [Comparison]
+      // Custom Algorithm:
+      // - square: 5, mul: 4, mul by non-residue: 3, add: 8, sub: 4, double: 4
+      // Default Karatsuba:
+      // - square: 4, mul: 6, mul by non-residue: 3, add: 6, sub: 0, double: 6
+      //
+      // Conclusion:
+      // The Custom algorithm saves 1 Multiplication at the cost of ~4
+      // Add/Sub. Since Mul >> Add in cost, this optimization is generally
+      // faster.
+      //
+      // For x = x₀ + x₁·u + x₂·u² + x₃·u³ where u⁴ = ξ:
+      //
+      // s₀ = x₀²
+      // s₁ = x₁²
+      // s₂ = x₂²
+      // s₃ = x₃²
+      // s₄ = (x₀ + x₂)²
+      // m₀ = 2 * x₀ * x₁
+      // m₁ = 2 * x₂ * x₃
+      // m₂ = 2 * x₁ * x₃
+      // m₃ = 2 * (x₀ + x₂) * (x₁ + x₃)
+      //
+      // Result:
+      // y₀ = s₀ + ξ * (s₂ + 2x₁x₃)
+      // y₁ = m₀ + ξ * m₁
+      // y₂ = s₁ + ξ * (s₄ - s₀ - s₂)
+      // y₃ = m₃ - m₀ - m₁
+
+      std::array<BaseField, 4> x =
+          static_cast<const Derived&>(*this).ToBaseField();
+      BaseField non_residue = static_cast<const Derived&>(*this).NonResidue();
+
+      // 1. Squares (xᵢ²)
+      // s₀ = x₀²
+      BaseField s0 = x[0].Square();
+      // s₁ = x₁²
+      BaseField s1 = x[1].Square();
+      // s₂ = x₂²
+      BaseField s2 = x[2].Square();
+      // s₃ = x₃²
+      BaseField s3 = x[3].Square();
+
+      // s₄ = (x₀ + x₂)²
+      BaseField s4 = (x[0] + x[2]).Square();
```

Comment:
We can reuse this(`x[0] + x[2]`) in the line 126(`((x[0] + x[2]) * (x[1] + x[3])).Double();`)

```
  BaseField x0_plus_x2 = x[0] + x[2];                                                     
  BaseField s4 = x0_plus_x2.Square();                                                     
  // ...                                                                                  
  BaseField m3 = (x0_plus_x2 * (x[1] + x[3])).Double(); 
  ```